go/src/koding/socialapi/cache.go

```diff
@@ -0,0 +1,165 @@
+package socialapi
+
+import (
+	"errors"
+	"sync"
+)
+
+// ErrSessionNotFound is returned by Storage.Get
+// when requested session was not found.
+var ErrSessionNotFound = errors.New("session not found")
+
+// Storage is an interface for external session storage.
+type Storage interface {
+	// Get gets a session.
+	//
+	// It returns ErrSessionNotFound if requested
+	// session was not cached.
+	Get(*Session) error
+
+	// Set sets a session.
+	//
+	// It overwrites any session that may exist
+	// already.
+	Set(*Session) error
+
+	// Delete deletes a session.
+	//
+	// If requested session does not exist,
+	// the method is a nop.
+	Delete(*Session) error
+}
+
+// SessionCache is a wrapper for AuthFunc that caches
+// sessions and invalidates them when requested.
+type SessionCache struct {
+	// AuthFunc is a mean to obtain session information.
+	//
+	// It is only called when a session was requested
+	// for a user/team pair that is not cached yet.
+	// Or when we were explicitly asked to request
+	// new session with Refresh set to true.
+	AuthFunc AuthFunc
+
+	// Storage is an external storage for storing session.
+	//
+	// If nil, in-memory map is going to be used instead.
+	Storage Storage
+
+	cacheMu sync.RWMutex
+	cache   map[string]*Session
```

Comment:
> lets use kodig/cache whenever possible

The `cache` is default memory cache if Storage is nil. I avoided koding/cache (the `Memory` one) since I did not like it implementation [0] - moreover the implementation is so trivial (map + mutex) that I thought it does not varrant pulling in external dependency. 

[0] should have sync.RWMutex instead - https://github.com/koding/cache/blob/master/memory.go